src/lib.rs

```diff
@@ -2341,6 +2341,26 @@ impl Connection {
 
                     self.peer_transport_params = peer_params;
 
+                    // Update max_data of current sender streams with some data
+                    // already written. This can happen when there is a
+                    // Handshake packet loss but received actual request already,
+                    // so the stream (usually id 0) is already opened and have some
+                    // data to write, but actual packetization is prevented because of
+                    // max_tx_data = 0.
+                    // What we do here is update max_data if max_data=0
+                    // In normal situation (handshake is complete before getting
+                    // request) it will not do anything.
+                    for (id, stream) in self.streams.iter_mut() {
+                        if stream.send.get_max_data() == 0 {
+                            stream.send.update_max_data(self.max_tx_data);
```

Comment:
`self.max_tx_data` is the connection-wide `peer_params.initial_max_data`, so I think this inadvertently sets the stream credit window to a size that is generally larger. In the case of quiche examples, there is an order of magnitude difference between the values. We should probably use `peer_params.initial_max_stream_data_bidi_remote` (which is what stream.rs does internally when it creates the stream object).

Reply:
`self.max_tx_data` is initialized when handshake completes and will be used when a stream is created by Stream::new(). When stream is created (using `get_or_create()`) it will use `peer_params.initial_max_stream_data_bidi_local`, not `remote`. So I don't need to use peer_params value explicitly here.

Reply:
I don't follow. On the server side, the stream is created on reception of a Stream frame (https://github.com/cloudflare/quiche/blob/master/src/lib.rs#L1362). It seems the problem is triggered when the server receives STREAM frame before do_handshake() returns Ok?

If thats the case then the stream is constructed from default transport paramers of 0 stream window size. In the linked code:

```
let max_tx_data = self
                        .peer_transport_params
                        .initial_max_stream_data_bidi_local
                        as usize;
```
that value propagates down to into the stream constructor

```
Stream {
            recv: RecvBuf::new(max_rx_data),
            send: SendBuf::new(max_tx_data),
        }
```

But with this PR, we read the peer's initial max data into the Connection (self) and then use that value to update the stream's send buff tx size. That can't be right unless I'm missing something.

```
self.max_tx_data = peer_params.initial_max_data as usize;
...
stream.send.update_max_data(self.max_tx_data);
```

Reply:
Got it. So I was somehow confused with Connection.max_data and Stream.send.max_data. We need to use initial_max_stream_data_bidi_local. I will update.

Reply:
Fixed in 6fe07a4